p2p/metrics.go

```diff
@@ -19,53 +19,200 @@
+	id           string        // NodeID of the peer
+	metered      bool          // Checks if the peer is metered
+	ingressMeter metrics.Meter // Meter for the read bytes of the peer
+	egressMeter  metrics.Meter // Meter for the written bytes of the peer
+
+	lock sync.RWMutex // Lock protecting the metered connection's internals
 }
 
-// newMeteredConn creates a new metered connection, also bumping the ingress or
-// egress connection meter. If the metrics system is disabled, this function
-// returns the original object.
-func newMeteredConn(conn net.Conn, ingress bool) net.Conn {
+// newMeteredConn creates a new metered connection, bumps the ingress or egress
+// connection meter and also increases the metered peer count. If the metrics
+// system is disabled or the IP address is unspecified, this function returns
+// the original object.
+func newMeteredConn(conn net.Conn, ingress bool, ip net.IP) net.Conn {
 	// Short circuit if metrics are disabled
 	if !metrics.Enabled {
 		return conn
 	}
-	// Otherwise bump the connection counters and wrap the connection
+	if ip.IsUnspecified() {
+		log.Warn("Peer IP is unspecified")
+		return conn
+	}
+	// Bump the connection counters and wrap the connection
 	if ingress {
 		ingressConnectMeter.Mark(1)
 	} else {
 		egressConnectMeter.Mark(1)
 	}
-	return &meteredConn{Conn: conn}
+	return &meteredConn{
+		Conn:      conn,
+		ip:        ip,
+		connected: time.Now(),
+	}
 }
 
-// Read delegates a network read to the underlying connection, bumping the ingress
-// traffic meter along the way.
+// Read delegates a network read to the underlying connection, bumping the common
+// and the peer ingress traffic meters along the way.
 func (c *meteredConn) Read(b []byte) (n int, err error) {
 	n, err = c.Conn.Read(b)
 	ingressTrafficMeter.Mark(int64(n))
-	return
+	c.lock.RLock()
+	if c.metered {
+		c.ingressMeter.Mark(int64(n))
+	}
+	c.lock.RUnlock()
+	return n, err
 }
 
-// Write delegates a network write to the underlying connection, bumping the
-// egress traffic meter along the way.
+// Write delegates a network write to the underlying connection, bumping the common
+// and the peer egress traffic meters along the way.
 func (c *meteredConn) Write(b []byte) (n int, err error) {
 	n, err = c.Conn.Write(b)
 	egressTrafficMeter.Mark(int64(n))
-	return
+	c.lock.RLock()
+	if c.metered {
+		c.egressMeter.Mark(int64(n))
+	}
+	c.lock.RUnlock()
+	return n, err
+}
+
+// handshakeDone is called when a peer handshake is done. Registers the peer to
+// the ingress and the egress traffic registries using the peer's IP and NodeID
+// if the metered peer count didn't reach the limit, also emits connect event.
+func (c *meteredConn) handshakeDone(nodeID discover.NodeID) {
+	if atomic.LoadInt32(&meteredPeerCount) >= MeteredPeerLimit {
+		log.Warn("Metered peer count reached the limit")
+		return
+	}
+	// Increment the metered peer count
+	atomic.AddInt32(&meteredPeerCount, 1)
```

Comment:
This construct is racey, because you check first and increment afterwards, but it could change in between.

```go
if atomic.LoadInt32(&meteredPeerCount) >= MeteredPeerLimit {
	return
}
atomic.AddInt32(&meteredPeerCount, 1)
```

If you want to ensure that we stay correctly below `MeteredPeerLimit`, you atomically add-and-retrieve and then see if you went below the allowance. If yes, you can decrement back. A bit more work, but correct AFAIK:

```go
if atomic.AddInt32(&meteredPeerCount, 1) >= MeteredPeerLimit {
	atomic.AddInt32(&meteredPeerCount, -1)
	return
}
```

@fjl Could you please confirm this? ^